File: scripts/check-packaged-assets.js
#!/usr/bin/env node

const fs = require("fs");
const os = require("os");
const path = require("path");
const childProcess = require("child_process");

const usage = `Usage: scripts/check-packaged-assets.js <plugin.vcvplugin|plugin-dir|artifact-dir> [...]`;

const repoRoot = path.resolve(__dirname, "..");
const sourceExtensions = new Set([
  ".c",
  ".cc",
  ".cpp",
  ".h",
  ".hpp",
  ".hh",
  ".json",
  ".md",
]);

function walk(dir, files = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (
      entry.name === ".git" ||
      entry.name === "build" ||
      entry.name === "dist" ||
      entry.name === "node_modules"
    ) {
      continue;
    }

    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walk(fullPath, files);
    } else {
      files.push(fullPath);
    }
  }
  return files;
}

function sourceFiles() {
  return walk(repoRoot).filter((file) => {
    const rel = path.relative(repoRoot, file);
    if (rel.startsWith(`res${path.sep}`)) return false;
    if (rel.startsWith(`scripts${path.sep}gen${path.sep}`)) return false;
    if (path.basename(file) === "Makefile") return true;
    return sourceExtensions.has(path.extname(file));
  });
}

function referencedAssets() {
  const refs = new Map();
  const directPluginAsset =
    /asset::plugin\s*\([^;]*?["'`]((?:res\/)[^"'`)\s]+)["'`][^;]*?\)/gs;
  const resVariable =
    /(?:std::string|const\s+char\s*\*|char\s+const\s*\*)\s+([A-Za-z_][A-Za-z0-9_]*)\s*=\s*["'`]((?:res\/)[^"'`)\s]+)["'`]/g;

  for (const file of sourceFiles()) {
    const relFile = path.relative(repoRoot, file);
    const contents = fs.readFileSync(file, "utf8");
    let match;

    while ((match = directPluginAsset.exec(contents))) {
      const asset = match[1];
      if (!refs.has(asset)) refs.set(asset, []);
      refs.get(asset).push(relFile);
    }

    while ((match = resVariable.exec(contents))) {
      const [, variable, asset] = match;
      const pluginVariable = new RegExp(
        `asset::plugin\\s*\\([^;]*\\b${variable}\\b[^;]*\\)`,
        "s",
      );
      if (!pluginVariable.test(contents)) continue;

      if (!refs.has(asset)) refs.set(asset, []);
      refs.get(asset).push(relFile);
    }
  }

  return refs;
}

function run(command, args, options = {}) {
  return childProcess.execFileSync(command, args, {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    ...options,
  });
}

function extractPackage(packagePath) {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "vcvplugin-assets-"));
  const tarPath = path.join(tmpDir, "plugin.tar");

  try {
    run("zstd", ["-d", packagePath, "-o", tarPath]);
    fs.mkdirSync(path.join(tmpDir, "unpacked"));
    run("tar", ["-xf", tarPath, "-C", path.join(tmpDir, "unpacked")]);
    return { root: path.join(tmpDir, "unpacked"), cleanup: () => rm(tmpDir) };
  } catch (error) {
    rm(tmpDir);
    throw error;
  }
}

function rm(target) {
  fs.rmSync(target, { recursive: true, force: true });
}

function findPackages(dir) {
  return walk(dir).filter((file) => file.endsWith(".vcvplugin"));
}

function findUnpackedPluginRoots(dir) {
  return walkDirs(dir).filter(
    (candidate) =>
      fs.existsSync(path.join(candidate, "plugin.json")) &&
      fs.existsSync(path.join(candidate, "res")),
  );
}

function walkDirs(dir, dirs = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (entry.name === ".git" || entry.name === "node_modules") continue;

    const fullPath = path.join(dir, entry.name);
    if (!entry.isDirectory()) continue;

    dirs.push(fullPath);
    walkDirs(fullPath, dirs);
  }
  return dirs;
}

function packageRoots(targetPath) {
  const stat = fs.statSync(targetPath);
  if (stat.isFile()) {
    if (!targetPath.endsWith(".vcvplugin")) {
      throw new Error(`Expected a .vcvplugin file: ${targetPath}`);
    }
    return [{ label: targetPath, ...extractPackage(targetPath) }];
  }

  if (!stat.isDirectory()) {
    throw new Error(`Expected a file or directory: ${targetPath}`);
  }

  if (
    fs.existsSync(path.join(targetPath, "plugin.json")) &&
    fs.existsSync(path.join(targetPath, "res"))
  ) {
    return [{ label: targetPath, root: targetPath, cleanup: () => {} }];
  }

  const packages = findPackages(targetPath);
  if (packages.length > 0) {
    return packages.map((packagePath) => ({
      label: packagePath,
      ...extractPackage(packagePath),
    }));
  }

  const unpackedRoots = findUnpackedPluginRoots(targetPath);
  if (unpackedRoots.length === 0) {
    throw new Error(
      `No .vcvplugin files or unpacked plugin root found in: ${targetPath}`,
    );
  }

  return unpackedRoots.map((root) => ({
    label: root,
    root,
    cleanup: () => {},
  }));
}

function packagedPaths(root) {
  const files = walk(root);
  const rootEntries = fs
    .readdirSync(root, { withFileTypes: true })
    .filter((entry) => entry.isDirectory());

  const pluginRoot =
    rootEntries.length === 1 &&
    fs.existsSync(path.join(root, rootEntries[0].name, "plugin.json"))
      ? path.join(root, rootEntries[0].name)
      : root;

  return new Set(
    files.map((file) => path.relative(pluginRoot, file).split(path.sep).join("/")),
  );
}

function validateOne(target, refs) {
  const files = packagedPaths(target.root);
  const missing = [];

  for (const [asset, locations] of refs) {
    if (!files.has(asset)) {
      missing.push({ asset, locations });
    }
  }

  if (missing.length === 0) {
    console.log(`OK: ${target.label}`);
    return 0;
  }

  console.error(`Missing packaged assets in ${target.label}:`);
  for (const miss of missing) {
    console.error(`  Missing packaged asset: ${miss.asset}`);
    console.error(`    referenced by: ${miss.locations.join(", ")}`);
  }
  return missing.length;
}

function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.error(usage);
    process.exit(2);
  }

  const refs = referencedAssets();
  if (refs.size === 0) {
    console.error("No quoted res/... references found.");
    process.exit(2);
  }

  let failures = 0;
  const roots = [];

  try {
    for (const arg of args) {
      roots.push(...packageRoots(path.resolve(arg)));
    }

    for (const root of roots) {
      failures += validateOne(root, refs);
    }
  } finally {
    for (const root of roots) {
      root.cleanup();
    }
  }

  if (failures > 0) process.exit(1);
}

main();

File: scripts/check-packaged-assets.py
#!/usr/bin/env python3

import os
import re
import shutil
import subprocess
import sys
import tarfile
import tempfile
from pathlib import Path


USAGE = (
    "Usage: scripts/check-packaged-assets.py "
    "<plugin.vcvplugin|plugin-dir|artifact-dir> [...]"
)

REPO_ROOT = Path(__file__).resolve().parent.parent
SOURCE_EXTENSIONS = {
    ".c",
    ".cc",
    ".cpp",
    ".h",
    ".hpp",
    ".hh",
    ".json",
    ".md",
}


def walk_files(root):
    ignored = {".git", "build", "dist", "node_modules"}
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [name for name in dirnames if name not in ignored]
        for filename in filenames:
            yield Path(dirpath) / filename


def source_files():
    for file_path in walk_files(REPO_ROOT):
        rel = file_path.relative_to(REPO_ROOT)
        if rel.parts and rel.parts[0] == "res":
            continue
        if len(rel.parts) >= 2 and rel.parts[0] == "scripts" and rel.parts[1] == "gen":
            continue
        if file_path.name == "Makefile" or file_path.suffix in SOURCE_EXTENSIONS:
            yield file_path


def add_ref(refs, asset, rel_file):
    refs.setdefault(asset, [])
    if rel_file not in refs[asset]:
        refs[asset].append(rel_file)


def referenced_assets():
    refs = {}
    direct_plugin_asset = re.compile(
        r"asset::plugin\s*\([^;]*?[\"'`]((?:res/)[^\"'`)\s]+)[\"'`][^;]*?\)",
        re.S,
    )
    res_variable = re.compile(
        r"(?:std::string|const\s+char\s*\*|char\s+const\s*\*)\s+"
        r"([A-Za-z_][A-Za-z0-9_]*)\s*=\s*"
        r"[\"'`]((?:res/)[^\"'`)\s]+)[\"'`]"
    )

    for file_path in source_files():
        rel_file = file_path.relative_to(REPO_ROOT).as_posix()
        contents = file_path.read_text(encoding="utf-8")

        for match in direct_plugin_asset.finditer(contents):
            add_ref(refs, match.group(1), rel_file)

        for variable, asset in res_variable.findall(contents):
            plugin_variable = re.compile(
                rf"asset::plugin\s*\([^;]*\b{re.escape(variable)}\b[^;]*\)",
                re.S,
            )
            if plugin_variable.search(contents):
                add_ref(refs, asset, rel_file)

    return refs


def run(command):
    subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)


def extract_package(package_path):
    temp_dir = Path(tempfile.mkdtemp(prefix="vcvplugin-assets-"))
    tar_path = temp_dir / "plugin.tar"
    unpacked = temp_dir / "unpacked"

    try:
        run(["zstd", "-d", str(package_path), "-o", str(tar_path)])
        unpacked.mkdir()
        with tarfile.open(tar_path) as tar:
            tar.extractall(unpacked)
        return {
            "label": str(package_path),
            "root": unpacked,
            "cleanup": lambda: shutil.rmtree(temp_dir, ignore_errors=True),
        }
    except Exception:
        shutil.rmtree(temp_dir, ignore_errors=True)
        raise


def walk_dirs(root):
    for dirpath, dirnames, _filenames in os.walk(root):
        dirnames[:] = [name for name in dirnames if name not in {".git", "node_modules"}]
        for dirname in dirnames:
            yield Path(dirpath) / dirname


def find_packages(root):
    return [file_path for file_path in walk_files(root) if file_path.suffix == ".vcvplugin"]


def is_unpacked_plugin_root(root):
    return (root / "plugin.json").exists() and (root / "res").exists()


def find_unpacked_plugin_roots(root):
    return [candidate for candidate in walk_dirs(root) if is_unpacked_plugin_root(candidate)]


def package_roots(target_path):
    if target_path.is_file():
        if target_path.suffix != ".vcvplugin":
            raise ValueError(f"Expected a .vcvplugin file: {target_path}")
        return [extract_package(target_path)]

    if not target_path.is_dir():
        raise ValueError(f"Expected a file or directory: {target_path}")

    if is_unpacked_plugin_root(target_path):
        return [{"label": str(target_path), "root": target_path, "cleanup": lambda: None}]

    packages = find_packages(target_path)
    if packages:
        return [extract_package(package_path) for package_path in packages]

    unpacked_roots = find_unpacked_plugin_roots(target_path)
    if not unpacked_roots:
        raise ValueError(f"No .vcvplugin files or unpacked plugin root found in: {target_path}")

    return [
        {"label": str(root), "root": root, "cleanup": lambda: None}
        for root in unpacked_roots
    ]


def packaged_paths(root):
    root_entries = [entry for entry in root.iterdir() if entry.is_dir()]
    if len(root_entries) == 1 and (root_entries[0] / "plugin.json").exists():
        plugin_root = root_entries[0]
    else:
        plugin_root = root

    files = []
    for file_path in walk_files(plugin_root):
        try:
            files.append(file_path.relative_to(plugin_root).as_posix())
        except ValueError:
            continue

    return set(files)


def validate_one(target, refs):
    files = packaged_paths(target["root"])
    missing = []

    for asset, locations in refs.items():
        if asset not in files:
            missing.append((asset, locations))

    if not missing:
        print(f"OK: {target['label']}")
        return 0

    print(f"Missing packaged assets in {target['label']}:", file=sys.stderr)
    for asset, locations in missing:
        print(f"  Missing packaged asset: {asset}", file=sys.stderr)
        print(f"    referenced by: {', '.join(locations)}", file=sys.stderr)
    return len(missing)


def main():
    if len(sys.argv) < 2:
        print(USAGE, file=sys.stderr)
        return 2

    refs = referenced_assets()
    if not refs:
        print("No quoted res/... references found.", file=sys.stderr)
        return 2

    roots = []
    failures = 0
    try:
        for arg in sys.argv[1:]:
            roots.extend(package_roots(Path(arg).resolve()))

        for root in roots:
            failures += validate_one(root, refs)
    finally:
        for root in roots:
            root["cleanup"]()

    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())
